src/Compiler/Facilities/BuildGraph.fs

```diff
@@ -228,210 +223,84 @@ module GraphNode =
-                                            return! p
-                                        },
-                                        (fun res ->
-                                            cachedResult <- Task.FromResult(res)
-                                            cachedResultNode <- node.Return res
-                                            computation <- Unchecked.defaultof<_>
-                                            tcs.SetResult(res)),
-                                        (fun ex -> tcs.SetException(ex)),
-                                        (fun _ -> tcs.SetCanceled()),
-                                        ct
-                                    )
-
-                                    return! tcs.Task |> NodeCode.AwaitTask
-                            finally
-                                if taken then semaphore.Release() |> ignore
-                        finally
-                            lock gate <| fun () -> requestCount <- requestCount - 1
-
-                    | GraphNodeAction.GetValueByAgent ->
-                        assert (not retryCompute)
-                        let mbp, cts = agent
-
-                        try
-                            let! ct = NodeCode.CancellationToken
-
-                            let! res =
-                                mbp.PostAndAsyncReply(fun replyChannel -> GetValue(replyChannel, ct))
-                                |> NodeCode.AwaitAsync
-
-                            match res with
-                            | Ok result -> return result
-                            | Result.Error ex -> return raise ex
-                        finally
-                            lock gate
-                            <| fun () ->
-                                requestCount <- requestCount - 1
-
-                                if requestCount = 0 then
-                                    cts.Cancel() // cancel computation when all requests are cancelled
-
-                                    try
-                                        (mbp :> IDisposable).Dispose()
-                                    with _ ->
-                                        ()
-
-                                    cts.Dispose()
-                                    agent <- Unchecked.defaultof<_>
+                Interlocked.Increment(&requestCount) |> ignore
+                try
+                    let! ct = NodeCode.CancellationToken
+
+                    // We must set 'taken' before any implicit cancellation checks
+                    // occur, making sure we are under the protection of the 'try'.
+                    // For example, NodeCode's 'try/finally' (TryFinally) uses async.TryFinally which does
+                    // implicit cancellation checks even before the try is entered, as do the
+                    // de-sugaring of 'do!' and other NodeCode constructs.
+                    let mutable taken = false
+
+                    try
+                        do!
```

Comment:
While we wait for the semaphore the computation could be running in a different thread (and most likely it is, if we have to wait). Then when we get the semaphore the result is now available.

There were actually more checks before which I removed. Hopefully they are not needed anymore.

Reply:
Well, I thought that If cachedResult is there then we might first check it. Because if computation is done and cached already then taking a semaphore is presumably not really needed. But I might be wrong. 

I tried to switch NodeCode to Task for better stacks and perf but it turned out faulty (but with better stacks :D)

Your changes most probably will make that transition easier (well if you ever decide to not use async as nodeCode engine)

Reply:
Well we do check that first on line 241. If there is `cachedResultNode`, it's always `node.Return res` (from line 281) with result already computed.

Reply:
It might make sense to check again if there can be a situation when someone would get the `NodeCode` and not evaluate it immediately. But I don't think we have such scenarios.